vm/src/pyobject.rs

```diff
@@ -926,6 +927,12 @@ pub trait PyValue: fmt::Debug + PyThreadingConstraint + Sized + 'static {
         self.into_ref(vm).into_object()
     }
 
+    fn special_retrieve(vm: &VirtualMachine, obj: PyObjectRef) -> Option<PyResult<PyRef<Self>>> {
```

Comment:
I think this could be `&PyObjectRef`, since `try_from_object` always calls `special_retrieve` with `obj.clone()`. Also I think you can put `#[inline(always)]`